### PR TITLE
Add optional systemd unit

### PR DESCRIPTION
**Pull request by terencode**
~~While it might not seem useful at first, it will be if it ships with an optional systemd unit file because the user will just have to change the options and enable the unit.~~

### CODE REVIEW

**Comment by terencode** [comment]
Here is an example PKGBUILD taking this MR into account:
The original behaviour is preserved but now the user can set up its config and enable the systemd user unit.
```
pkgname=jack-matchmaker-git
pkgver=0.7.1.r7.g519f280
pkgrel=1
pkgdesc="A command line tool to auto-connect JACK ports matching given patterns. Git version."
arch=('any')
url="https://github.com/SpotlightKid/jack-matchmaker"
license=('GPL2')
depends=('python-setuptools' 'jack')
conflicts=('jack-matchmaker')
source=("jack-matchmaker-git::git+https://github.com/terencode/jack-matchmaker.git")
sha256sums=('SKIP')
pkgver() {
  cd "${srcdir}/${pkgname}"
  git describe --long --tags | sed 's/\([^-]*-g\)/r\1/;s/-/./g'
}


build() {
  cd "${srcdir}/${pkgname}"

  python setup.py build
}

package() {
  cd "${srcdir}/${pkgname}"

  python setup.py install --skip-build --root="${pkgdir}" --prefix=/usr \
    --optimize=1

  install -Dm 644 example.conf $pkgdir/etc/jack-matchmaker/jack-matchmaker.conf.default
  install -Dm 644 jack-matchmaker.service -t $pkgdir/usr/lib/systemd/user/
}
```

File: jackmatchmaker/__init__.py [hunk 1]
```diff
@@ -169,6 +186,10 @@ def reg_callback(self, port_id, action, *args):
         log.debug("New port registered: %s", jacklib.port_name(port))
         self._refresh()
 
+    def error_callback(self, error):
+        error = error.decode('utf-8')
+        log.info(error)
+
```
Comment (SpotlightKid) [inline jackmatchmaker/__init__.py]: This change, along the with changes to `jacklib.py`, should go into a separate PR.

File: jackmatchmaker/jacklib.py [hunk 2]
```diff
@@ -1262,6 +1263,15 @@ def get_time():
 jlib.jack_free.argtypes = [c_void_p]
 jlib.jack_free.restype = None
 
+jlib.jack_set_error_function.argtypes = [JackErrorCallback]
```
Comment (SpotlightKid) [inline jackmatchmaker/jacklib.py]: * `jack_set_error_function` is part of the optional API, so you need to wrap this in a try/except:

       try:
           jlib.jack_set_error_function.argtypes = ...
           jlib.jack_set_error_function.restype = None
       except AttributeError:
           jlib.jack_set_error_function = None

* `jack_set_error_function` returns `void`, so you need to add

       jlib.jack_set_error_function.restype = None

File: jackmatchmaker/jacklib.py [hunk 3]
```diff
@@ -1262,6 +1263,15 @@ def get_time():
...
+
+def set_error_callback(error_callback) :
+    if jlib.jack_set_error_function:
+        global _error_callback
+        _error_callback = JackErrorCallback(error_callback)
+        jlib.jack_set_error_function(_error_callback)
+
+    return -1
```
Comment (SpotlightKid) [inline jackmatchmaker/jacklib.py]: Please run 'flake8' on all changed Python files to check for coding-style errors, such as the missing empty line here.

File: jackmatchmaker/jacklib.py [hunk 4]
```diff
@@ -1262,6 +1263,15 @@ def get_time():
 jlib.jack_free.argtypes = [c_void_p]
 jlib.jack_free.restype = None
 
+jlib.jack_set_error_function.argtypes = [JackErrorCallback]
+
+def set_error_callback(error_callback) :
+    if jlib.jack_set_error_function:
+        global _error_callback
```
Comment (SpotlightKid) [inline jackmatchmaker/jacklib.py]: The `_error_callback` should be initialized to `None` at the module global namespace.

Also, please put `global _error_callback` at the start of the function body.

File: jackmatchmaker/__init__.py [hunk 5]
```diff
@@ -358,19 +381,37 @@ def main(args=None):
                     " (default: %(default)s)")
     ap.add_argument('-m', '--max-attempts', type=posnum, default=0, metavar="NUM",
                     help="Max. number of attempts to connect to JACK server (default: 0=infinite)")
-    ap.add_argument('-v', '--verbose', action="store_true", help="Be verbose")
-    ap.add_argument('--version', action='version', version='%%(prog)s %s' % __version__)
```
Comment (SpotlightKid) [inline jackmatchmaker/__init__.py]: Why did you remove the `--version` argument?

File: jackmatchmaker/__init__.py [hunk 6]
```diff
@@ -358,19 +381,37 @@ def main(args=None):
...
+    ap.add_argument('-v', '--verbose', nargs='?', metavar="NUM",
```
Comment (SpotlightKid) [inline jackmatchmaker/__init__.py]: Why `metavar="NUM"` if the choices are level *labels*?

File: jackmatchmaker/__init__.py [hunk 7]
```diff
@@ -358,19 +381,37 @@ def main(args=None):
...
+                    choices=['DEBUG', 'INFO', 'WARNING', 'ERROR'], const='DEBUG', default='INFO',
+                    help="Set verbosity level. (choices: %(choices)s, default: %(default)s). "
+                    "Passing no arguments sets it to DEBUG ")
```
Comment (SpotlightKid) [inline jackmatchmaker/__init__.py]: If this becomes a choice options, then the long form should probably be changed to `--verbosity` or `--log-level`.

Comment (SpotlightKid) [reply jackmatchmaker/__init__.py]: This is also an unrelated change and should go into a separate PR.

File: jackmatchmaker/__init__.py [hunk 8]
```diff
@@ -69,6 +81,8 @@ def __init__(self, patterns, pattern_file=None, name=__program__, exact_matching
         self.connect_interval = connect_interval
         self.default_encoding = jacklib.ENCODING
 
+        jacklib.set_error_callback(self.error_callback)
```
Comment (SpotlightKid) [inline jackmatchmaker/__init__.py]: This change, along the with changes to `jacklib.py` and the new `error_callback` method, should go into a separate PR.

File: jackmatchmaker/__init__.py [hunk 9]
```diff
@@ -358,19 +381,37 @@ def main(args=None):
...
+                            # if pattern file path is not absolute, use the path of the config file
+                            value = "'{}/{}'".format(path.dirname(args.config_file), value.strip("\""))
+                        exec("args.%s = %s" % (variable, value))
+                    else:
+                        log.error("Incorrect config option: '%s'", variable)
+        except FileNotFoundError:
+            sys.exit("ERROR: Config file '{}' not found".format(path.realpath(args.config_file)))
+
 
+    logging.basicConfig(level=logging.getLevelName(args.verbose), format="%(levelname)s: %(message)s")
```
Comment (SpotlightKid) [inline jackmatchmaker/__init__.py]: Since `args.verbose` is already guaranteed to be a valid log level name, and the `level` argument accepts a level name (since Python 3.2), the use of `logging.getLevelName` here makes no sense.

File: jackmatchmaker/__init__.py [hunk 10]
```diff
@@ -347,6 +368,8 @@ def main(args=None):
                      help="Include pretty-names from port meta data when listing ports")
     ap.add_argument('-p', '--pattern-file', metavar="FILE",
                     help="Read pattern pairs from FILE (one pattern per line)")
+    ap.add_argument('--config-file', metavar="FILE",
+                    help="Read program arguments from FILE")
```
Comment (SpotlightKid) [inline jackmatchmaker/__init__.py]: A short option form (maybe `-C`?) would  nice.

File: jackmatchmaker/__init__.py [hunk 11]
```diff
@@ -126,12 +140,15 @@ def add_patterns(self, ptn_output, ptn_input):
...
+            with open(filename) as fp:
+                stripfilter = (line.strip() for line in fp)
+                linefilter = (line for line in stripfilter if line and not line.startswith('#'))
 
-            for ptn_output, ptn_input in pairwise(linefilter):
-                self.add_patterns(ptn_output, ptn_input)
+                for ptn_output, ptn_input in pairwise(linefilter):
+                    self.add_patterns(ptn_output, ptn_input)
+        except FileNotFoundError:
+            sys.exit("ERROR: Pattern file '{}' not found".format(path.realpath(self.pattern_file)))
```
Comment (SpotlightKid) [inline jackmatchmaker/__init__.py]: The exception this function can raise should be caught where it is called (i.e in the `__init__` method). If the function can do nothing to recover from the error, it should not catch it. Certainly it should not `sys.exit` from the middle of an object method.

File: jackmatchmaker/__init__.py [hunk 12]
```diff
@@ -358,19 +381,37 @@ def main(args=None):
...
+                    if variable in vars(args) :
+                        # options directly passed to the program have the priority
+                        if variable == "pattern_file" and not value.count('/'):
+                            # if pattern file path is not absolute, use the path of the config file
+                            value = "'{}/{}'".format(path.dirname(args.config_file), value.strip("\""))
+                        exec("args.%s = %s" % (variable, value))
+                    else:
+                        log.error("Incorrect config option: '%s'", variable)
+        except FileNotFoundError:
+            sys.exit("ERROR: Config file '{}' not found".format(path.realpath(args.config_file)))
```
Comment (SpotlightKid) [inline jackmatchmaker/__init__.py]: Config file parsing is an independent functionality and should go in to its own function.

File: jackmatchmaker/__init__.py [hunk 13]
```diff
@@ -358,19 +381,37 @@ def main(args=None):
...
+            with open(args.config_file) as fp:
+                stripfilter = (line.strip() for line in fp)
+                linefilter = (line for line in stripfilter if line and not line.startswith('#'))
+                for variable, value in get_config_vars(linefilter):
+                    if variable in vars(args) :
+                        # options directly passed to the program have the priority
+                        if variable == "pattern_file" and not value.count('/'):
+                            # if pattern file path is not absolute, use the path of the config file
+                            value = "'{}/{}'".format(path.dirname(args.config_file), value.strip("\""))
+                        exec("args.%s = %s" % (variable, value))
```
Comment (SpotlightKid) [inline jackmatchmaker/__init__.py]: Don't use `exec` (or `eval`)  to parse use supplied config files *ever*.

File: jackmatchmaker/__init__.py [hunk 14]
```diff
@@ -44,6 +45,17 @@ def pairwise(iterable):
...
+def get_config_vars(iterable):
+    vars = []
+
+    for l in iterable :
+        l = l.replace(" ", "").split("=")
+        if (len(l) == 2):
+            vars.append(l[0].replace("-", "_"))
+            vars.append(l[1])
+
+    return pairwise(iter(vars))
```
Comment (SpotlightKid) [inline jackmatchmaker/__init__.py]: This is unnecessarily complicated. Why not just append `name, value`  as a tuple in the first place?

**Review by SpotlightKid — COMMENTED** [review]
Hi, I started this review and commented on a lot of details, but stopped at one point, because it became clear, that I'm not gonna accept this in the current form. I hope the existing comments are still helpful to you in some way.

**Comment by SpotlightKid** [comment]
First off, thanks for your work an your PR.

However, I'm not gonna accept this.

* There are too many unrelated changes in one PR:
  * jack errors function
  * verbosity command line switch
  * exception handling in `add_patterns_from_file`
  * config file handling
* The main change, the config file handling is not implemented in a way that meets my quality standards.
* But, most importantly, I am not convinced, that adding config file handling to jack-matchmaker is necessary.
 
  * You can either just add the necessary options to the `ExecStart` line in the systemd file
  * Or you can add a shell wrapper, that just reads the command line options by, for example, sourcing another shell file.

  I'd like to explore the latter possibility, before I would consider adding config file parsing directly to jack-matchmaker.

Your work was not in vain, though, the jack error function stuff is valuable, if you'd like to make a separate PR for that, I'll be more receptive.  Same goes for the verbosity switch. You also uncovered that exceptions raised by `add_pattern_from_file` were not caught everywhere (which I just fixed in master).

File: jackmatchmaker/__init__.py (hunk 5 above)
Comment (terencode) [reply jackmatchmaker/__init__.py]: Rebasing mistake...

File: jackmatchmaker/__init__.py (hunk 6 above)
Comment (terencode) [reply jackmatchmaker/__init__.py]: Forgot to update after changing it.

File: jackmatchmaker/__init__.py (hunk 7 above)
Comment (terencode) [reply jackmatchmaker/__init__.py]: I thought you'd be annoyed with too many PRs, that's why I combined everything in one...

**Comment by terencode** [comment]
Thanks for your detailed review.
I split the different change categories into different PR and applied your suggestions.
I gave the config parser a second thought and judged that it's indeed not needed If using the `EnvironmentFile` systemd variable.
I updated this PR to take this into account. It has a soft dependency on #9

File: jack-matchmaker.service [hunk 15]
```diff
@@ -0,0 +1,9 @@
+[Unit]
+Description=auto-connect JACK ports matching given patterns
+
+[Service]
+EnvironmentFile=/etc/conf.d/jack-matchmaker
+ExecStart=/usr/bin/jack-matchmaker -p $PATTERN_FILE -e $EXACT_MATCHING -N $CLIENT_NAME -m MAX_ATTEMPTS -v $VERBOSITY
```
Comment (SpotlightKid) [inline jack-matchmaker.service]: What syntax is supported here? Does this support Bash variable expansion syntax e.g.

     ${PATTERN_FILE+-p $PATTERN_FILE}

so you only pass command line options that are defined in the environment file?

Comment (SpotlightKid) [reply jack-matchmaker.service]: Also, do these placeholders need quoting for values with whitespace in them?

File: jack-matchmaker.service [hunk 16]
```diff
@@ -0,0 +1,9 @@
+[Unit]
+Description=auto-connect JACK ports matching given patterns
+
+[Service]
+EnvironmentFile=/etc/conf.d/jack-matchmaker
+ExecStart=/usr/bin/jack-matchmaker -p "$PATTERN_FILE" -e $EXACT_MATCHING -N "$CLIENT_NAME" -m $MAX_ATTEMPTS -v $VERBOSITY "$PATTERNS"
```
Comment (SpotlightKid) [inline jack-matchmaker.service]: The `-e` option does not take arguments, so `-e $EXACT_MATCHING` will not work as intended.

The only way to set this I see is to have something like `EXTRA_OPTIONS="-e"` in the EnvironmentFile and then add `$EXTRA_OPTIONS` to the command line (not quoted!).

Comment (SpotlightKid) [reply jack-matchmaker.service]: `$PATTERNS` must not be quoted, so that the patterns are passed as separate command line arguments.

See section "Command Lines" [here](https://www.freedesktop.org/software/systemd/man/systemd.service.html).

File: jack-matchmaker.conf [hunk 17]
```diff
@@ -0,0 +1,11 @@
+#
+# Parameters to be passed to jack-matchmaker
+#
+
+PATTERN_FILE="/etc/jack-matchmaker/patterns.txt"
+PATTERNS=""
+CLIENT_NAME="jack-matchmaker"
+CONNECT_INTERVAL=3
```
Comment (SpotlightKid) [inline jack-matchmaker.conf]: This is missing in the command line (option `-I`).

**Review by SpotlightKid — CHANGES_REQUESTED** [review]

File: jack-matchmaker.service (hunk 15 above)
Comment (terencode) [reply jack-matchmaker.service]: Parameter expansion works by passing the commands to `/bin/sh -c` and escaping the `$` with `$$` so systemd doesn't do its filtering.

Yes they do need quoting.

File: jack-matchmaker.service (hunk 16 above)
Comment (terencode) [reply jack-matchmaker.service]: The `-e` option is fixed with parameter expansion: If EXACT_MATCHING is set to nothing it will not be presented as an argument.

File: jack-matchmaker.conf (hunk 17 above)
Comment (terencode) [reply jack-matchmaker.conf]: Added.

File: jack-matchmaker.service [hunk 18]
```diff
@@ -0,0 +1,9 @@
+[Unit]
+Description=auto-connect JACK ports matching given patterns
+
+[Service]
+EnvironmentFile=/etc/conf.d/jack-matchmaker
+ExecStart=/bin/sh -c "/usr/bin/jack-matchmaker $${PATTERN_FILE+-p \"$PATTERN_FILE\"} $${EXACT_MATCHING:+-e} $${CLIENT_NAME+-N \"$CLIENT_NAME\"} $${CONNECT_INTERVAL+-I $CONNECT_INTERVAL} $${MAX_ATTEMPTS+-m $MAX_ATTEMPTS} $${VERBOSITY+-v $VERBOSITY} $${PATTERNS+$PATTERNS}"
```
Comment (SpotlightKid) [inline jack-matchmaker.service]: Uh, that's quite ugly, but it works :) Better to use `/bin/bash` instead of `/bin/sh`, though, since the extended variable expansion syntax is Bash-specific and some systems might not have `/bin/sh` linked to `/bin/bash`.

Comment (SpotlightKid) [reply jack-matchmaker.service]: If you use outer single quotes and just use `$$PATTERNS` (undefined variables just expand to an empty string), you can simplify the line slightly:

    ExecStart=/bin/bash -c '/usr/bin/jack-matchmaker $${PATTERN_FILE+-p "$PATTERN_FILE"} $${EXACT_MATCHING:+-e} $${CLIENT_NAME+-N "$CLIENT_NAME"} $${CONNECT_INTERVAL+-I $CONNECT_INTERVAL} $${MAX_ATTEMPTS+-m $MAX_ATTEMPTS} $${VERBOSITY+-v $VERBOSITY} $$PATTERNS'

**Review by SpotlightKid — CHANGES_REQUESTED** [review]

File: jack-matchmaker.service (hunk 18 above)
Comment (terencode) [reply jack-matchmaker.service]: Thanks, updated :)

Comment (terencode) [reply jack-matchmaker.service]: ~~Uh using single quotes doesn't seem to work.~~ Nevermind it works.

Comment (terencode) [reply jack-matchmaker.service]: Uh single quotes doesn't seem to work

**Comment by terencode** [comment]
Here is an updated PKGBUILD that takes the new changes into account:
```
# Maintainer: Christopher Arndt <aur -at- chrisarndt -dot- de>

pkgname=jack-matchmaker-git
pkgver=0.7.1.r18.gcb49638
pkgrel=1
pkgdesc="A command line tool to auto-connect JACK ports matching given patterns. Git version."
arch=('any')
url="https://github.com/SpotlightKid/jack-matchmaker"
license=('GPL2')
depends=('python-setuptools' 'jack')
conflicts=('jack-matchmaker')
source=("jack-matchmaker-git::git+$url.git")
sha256sums=('SKIP')
backup=('etc/conf.d/jack-matchmaker')

pkgver() {
  cd "${srcdir}/${pkgname}"
  git describe --long --tags | sed 's/\([^-]*-g\)/r\1/;s/-/./g'
}


build() {
  cd "${srcdir}/${pkgname}"

  python setup.py build
}

package() {
  cd "${srcdir}/${pkgname}"

  python setup.py install --skip-build --root="${pkgdir}" --prefix=/usr \
    --optimize=1

  install -Dm 644 jack-matchmaker.conf $pkgdir/etc/conf.d/jack-matchmaker
  install -Dm 644 jack-matchmaker.service -t $pkgdir/usr/lib/systemd/user/
}
```

**Review by SpotlightKid — APPROVED** [review]
